// app/terminal/input/model.js
var Processor = require('app/terminal/processor');
var Prompt = require('app/terminal/input/prompt/model');
var EventEmitter = require('events').EventEmitter;

//CAN'T use terminal here, since we are initializing this
// object WHILE creating the terminal singleton!
//var terminal = require('app/terminal/model');


function Model(output){
  this.prompt = new Prompt();
  this.events = new EventEmitter();
  this.terminalOutput = output;

  this.defaultProcessor = new Processor('default');
  this.currentProcessor = this.defaultProcessor;

  this.prompt.events.on('run', this.run.bind(this));

  //delegate
  this.registerCommand = this.defaultProcessor.registerCommand.bind(this.defaultProcessor);
}

Model.prototype.setChars = function(chars){
  this.prompt.setChars(chars);
};

Model.prototype.addChar = function(char){
  this.prompt.addChar(char);
};

Model.prototype.special = function(specialChar){
  this.prompt[specialChar]();
};

Model.prototype.interrupt = function(){
  this.currentProcessor = this.defaultProcessor;
};

Model.prototype.run = function(inputString){
  var options = {
    enqueue: false,
    data: { promptName: this.currentProcessor.name}
  };

  this.terminalOutput.input(inputString, options);

  try {
    process.call(this, inputString);
  }
  catch(e) {
    this.terminalOutput.warn(e.message);
    endProcess.call(this);
    throw e;
  }
};

Model.prototype.getPromptName = function(){
  return this.currentProcessor.name;
};
// private, must be bound

// A Processor must return another Processor or
// a falsey value.  False indicates that the processor
// has terminated.  There is no stack, so a Processor
// cannot have a subProcessor.

var process = function(input) {
  this.events.emit('disabled');

  var calledClose = false;
  var self = this;

  this.currentProcessor.process(input).always(endProcess.bind(this));
};

var endProcess = function(newProcessor){

  this.currentProcessor = newProcessor || this.defaultProcessor;
  var events = this.events;
  this.terminalOutput.enqueue(
    events.emit.bind(events, 'enabled')
  );
};

module.exports = Model;
