// app/terminal/model.js
EventEmitter = require('events').EventEmitter;
Processor = require('app/terminal/processor');
Queue = require('app/utils/work_queue');

function Model(){
  this.events = new EventEmitter();
  this.queue = new Queue(10);
  this.lines = [];

  this.win = add.bind(this, 'win');
  this.log = add.bind(this, 'log');
  this.warn = add.bind(this, 'warn');

  this.defaultProcessor = new Processor();
  this.currentProcessor = this.defaultProcessor;
}

Model.prototype.input = function(input) {
  add.call(this, 'input', input, {enqueue: false});

  try {
    process.call(this, input);
  }
  catch(e) {
    this.warn(e.message);
    this.queue.push(this.events.emit.bind(this.events, 'processEnded'));
  }
};

Model.prototype.setInput = function(input) {
  this.events.emit('setInput', input);
};

Model.prototype.run = function(){
  this.events.emit('run');
};

Model.prototype.interrupt = function(){
  this.currentProcessor = this.defaultProcessor;
};

//private, must be bound

var add = function(type, text, options) {
  //options: enqueue: boolean

  var line = {
    type: type,
    text: text,
  };

  var displayLine = function(){
    this.lines.push(line);
    this.events.emit('lineAdded', line);
  };

  if (options && !options.enqueue){
    displayLine.call(this);
  }
  else {
    this.queue.push(displayLine.bind(this));
  }
};

// A Processor must return another Processor or
// a falsey value.  False indicates that the processor
// has terminated.  There is no stack, so a Processor
// cannot have a subProcessor.

var process = function(input) {
  this.events.emit('processStarted');
  this.currentProcessor = this.currentProcessor.process(input) || this.defaultProcessor;
  this.queue.push(this.events.emit.bind(this.events, 'processEnded'));
};

// singleton, because it will be easier
// than trying to pass the context around
// through asyncronous event handlers (e.g.,
// on url change...)

module.exports = new Model();
