// app/mobile/model.js
"use strict";

var EventEmitter = require('events').EventEmitter;

var states = [
  'confirmed mobile', // The user has taken an action confirming that they are using a mobile device
  'probably mobile', // The user has not taken an action to confirm, but our detector says that they are on mobile
  'probably not mobile', // The user has not taken an action to confirm, but our detector says that they are not on mobile
  'denied mobile', // The user has taken an action confirming that they are not using a mobile device
];

function Model(mobileDetected){
  this.events = new EventEmitter();
  if (mobileDetected){
    this.setState('probably mobile');
  }
  else {
    this.setState('probably not mobile');
  }
}

Model.prototype.setState = function(state){
  if (states.indexOf(state) === -1){
    throw Error("Invalid mobile detection state");
  }

  this.state = state;
  this.events.emit('stateChanged', this.state);
};

Model.prototype.getState = function(){
  return this.state;
};

// Other objects may need to know whether to act as mobile
// or not. This represents our best guess. Objects needing
// more granularity can ask for the state directly.
Model.prototype.isMobile = function(){
  return (states.indexOf(this.state) < 2);
};

module.exports = Model;
