// app/router/pushstate/router.js
"use strict";
var Parser = require('app/router/pushstate/parser');
var terminal = require('app/terminal/model');

var environment = 'local';
var resources = [
  'pages',
];


function Router(controllers){
  this.parser = new Parser(environment, resources);

  this.rootResource = controllers.root;
  delete controllers.root;

  this.controllers = controllers;
  window.onpopstate = this.routeEvent.bind(this);

  this.route(window.location.href);
}

Router.prototype.routeEvent = function(event){
  this.route(window.location.href);
};

Router.prototype.route = function(urlOrPath, routingDueToError){
  try {
    var resource = this.parser.parse(urlOrPath);

    if (resource === "root") {
      resource = this.rootResource;
    }
    this.controllers[resource.controller].load(resource.id);
  }
  catch(e) {
    if (routingDueToError){
      throw Error('infinite routing loop!');
    }

    console.log(e);
    this.push(this.rootResource, true);
    terminal.log('Routing to default page');
  }

};

Router.prototype.push = function(resource){
  var path = this.parser.construct(resource);
  window.history.pushState(null, null, path);
  this.route(path);
};

module.exports = Router;
