// app/terminal/output/model.js
'use strict';

var WorkQueue = require('app/utils/work_queue');
var EventEmitter = require('events').EventEmitter;
var Deferred = require('jquery').Deferred;

function Model(isOpen){
  this.workQueue = new WorkQueue(getDelay.bind(this, isOpen));
  this.lines = [];
  this.events = new EventEmitter();

  this.win = add.bind(this, 'win');
  this.log = add.bind(this, 'log');
  this.warn = add.bind(this, 'warn');
  this.input = add.bind(this, 'input');
}

Model.prototype.enqueue = function(fn){
  this.workQueue.push(fn);
};

//private, must be bound

//
//options:
//  enqueue: boolean
//  data: options for display
//
var add = function(type, text, options) {
  options = (options || {});

  var deferred = new Deferred();

  var line = {
    type: type,
    text: text,
    data: (options.data || {})
  };

  var displayLine = function(){
    this.lines.push(line);
    this.events.emit('lineAdded', line);
    deferred.resolve();
  };

  if (options.enqueue === false){
    displayLine.call(this);
  }
  else {
    this.workQueue.push(displayLine.bind(this));
  }

  return deferred.promise();
};

var getDelay = function(isOpen){
  if (isOpen()){
    return 200;
  }
  else {
    return 800;
  }
};

module.exports = Model;
